core/job/service/job_service.go

```diff
@@ -523,47 +546,40 @@ func (j JobService) generateJob(ctx context.Context, tenantWithDetails *tenant.W
 	return job.NewJob(tenantWithDetails.ToTenant(), spec, destination, sources), nil
 }
 
-func (j JobService) validateCyclic(rootName job.Name, jobMap map[job.Name]*job.Job, destinationToJobMap map[job.ResourceURN][]*job.Job) error {
-	dagTree, err := j.buildDAGTree(rootName, jobMap, destinationToJobMap)
-	if err != nil {
-		return err
-	}
-
+func (j JobService) validateCyclic(rootName job.Name, jobMap map[job.Name]*job.WithUpstream, destinationToJobMap map[job.ResourceURN][]*job.WithUpstream) ([]string, error) {
+	dagTree := j.buildDAGTree(rootName, jobMap, destinationToJobMap)
 	return dagTree.ValidateCyclic()
 }
 
-func (JobService) buildDAGTree(rootName job.Name, jobMap map[job.Name]*job.Job, destinationToJobMap map[job.ResourceURN][]*job.Job) (*tree.MultiRootTree, error) {
-	rootJob, ok := jobMap[rootName]
-	if !ok {
-		return nil, fmt.Errorf("couldn't find any job with name %s", rootName)
-	}
+func (JobService) buildDAGTree(rootName job.Name, jobMap map[job.Name]*job.WithUpstream, destinationToJobMap map[job.ResourceURN][]*job.WithUpstream) *tree.MultiRootTree {
+	rootJob, _ := jobMap[rootName]
 
 	dagTree := tree.NewMultiRootTree()
 	dagTree.AddNode(tree.NewTreeNode(rootJob))
 
 	for _, childJob := range jobMap {
 		childNode := findOrCreateDAGNode(dagTree, childJob)
-		for _, source := range childJob.Sources() {
-			if _, ok := destinationToJobMap[source]; !ok {
-				// source maybe from external optimus or outside project,
+		for _, upstream := range childJob.Upstreams() {
+			if _, ok := destinationToJobMap[upstream.Resource()]; !ok {
```

Comment:
Will this support jobs with static upstream that does not have destination/resource? as not all jobs have resource, for example jobs with python task

Reply:
Good point, in a current implementation, we use the resource as an identifier. In that case, if the job doesn't have a destination resource, then it won't be included in a cyclic deps detection.
Just thinking to use job name as identifier instead

Reply:
we might need to have both map, as for the unresolved inferred ones, the upstream won't have the job name yet.

Reply:
on second thought, I think having unresolved upstream might resulted in having a false negative case, where the cyclic is actually not there (as the upstream sensor will not be created if it is unresolved) for some cases. we can discuss more on the approach.